browser/extensions/brave_component_loader_browsertest.cc

```diff
@@ -0,0 +1,363 @@
+    if (added_ || removed_)
+      return true;
+    if (timeout_secs) {
+      timeout_ = timeout_secs;
+      elapsed_ = 0;
+      timer_.Start(FROM_HERE, base::TimeDelta::FromSeconds(1), this,
+                   &PdfjsExtensionExtensionRegistryObserver::OnTimer);
+    }
+    message_loop_runner_->Run();
+    return (added_ || removed_);
+  }
+
+  void OnTimer() {
+    if (++elapsed_ >= timeout_ || added_ || removed_)
+      timer_.Stop();
+    if (message_loop_runner_->loop_running())
+      base::ThreadTaskRunnerHandle::Get()->PostTask(
+          FROM_HERE, message_loop_runner_->QuitClosure());
+  }
+
+  void Reset() {
+    added_ = false;
+    removed_ = false;
+  }
+
+  // ExtensionRegistryObserver implementation:
+  void OnExtensionLoaded(content::BrowserContext* browser_context,
+    const extensions::Extension* extension) override {
+    ASSERT_EQ(extension->id(), kTestPdfjsExtensionId);
+    CHECK(!added_);
+    added_ = true;
+    if (added_ && message_loop_runner_->loop_running()) {
+      base::ThreadTaskRunnerHandle::Get()->PostTask(
+        FROM_HERE, message_loop_runner_->QuitClosure());
+    }
+  }
+
+  void OnExtensionUnloaded(
+    content::BrowserContext* browser_context,
+    const extensions::Extension* extension,
+    extensions::UnloadedExtensionReason reason) override {
+    ASSERT_EQ(extension->id(), kTestPdfjsExtensionId);
+    CHECK(!removed_);
+    removed_ = true;
+    if (removed_ && message_loop_runner_->loop_running()) {
+      base::ThreadTaskRunnerHandle::Get()->PostTask(
+        FROM_HERE, message_loop_runner_->QuitClosure());
+    }
+  }
+
+private:
+  scoped_refptr<content::MessageLoopRunner> message_loop_runner_;
+  bool added_;
+  bool removed_;
+  int timeout_;
+  int elapsed_;
+  base::RepeatingTimer timer_;
+  extensions::ExtensionRegistry* registry_;
+};
+
+} //namespace
+
+class BravePDFExtensionTest : public ExtensionFunctionalTest {
+public:
+  BravePDFExtensionTest() {
+    // Configure BraveComponentLoader to use test pdfjs.
+    BraveComponentLoader::SetPdfjsExtensionIdAndBase64PublicKeyForTest(
+      kTestPdfjsExtensionId, kTestPdfjsExtensionBase64PublicKey);
+  }
+  ~BravePDFExtensionTest() override = default;
+
+protected:
+  bool ShouldEnableInstallVerification() override {
+    // Simulates behavior in chrome/browser/extensions/install_verifier.cc's
+    // GetExperimentStatus.
+#if (defined(OS_WIN) || defined(OS_MACOSX))
+    return true;
+#else
+    return false;
+#endif
+  }
+
+  void SetDownloadPDFs(bool value) {
+    DCHECK(browser());
+    browser()->profile()->GetPrefs()->SetBoolean(
+      prefs::kPluginsAlwaysOpenPdfExternally, value);
+  }
+
+  void LoadPDFJSComponent(int timeout = 0, bool expect_loaded = true) {
+    // In browser test BraveComponentLoader doesn't actually load PDFJS
+    // component, so instead we unzip it where the component loader would
+    // expect to find the component and the toggle the preference to load it.
+    base::FilePath test_data_dir;
```

Comment:
We can't just use `InstallExtension` here to accomplish this?

Reply:
As we discussed, InstallExtension won't do here since the extension won't be registered as a component. LoadExtensionAsComponent would help with the initial load of the extension for the test. However, it would only be useful for testing the toggle in one direction (View PDFs in Browser -> Download PDFs) because in this case the code being tested would unload the extension. Testing in the opposite direction would not work, though, because that code still expects to find the component in the predetermined location.